### PR TITLE
fix GRASS7 r_stats_quantile

### PR DESCRIPTION
**Pull request by gioman**
## Description
This fixes 
https://issues.qgis.org/issues/20160

that is about the "r.stats.quantile.out" module (wrong output parameter name).

It also "fixes" the "r.stats.quantile.rast" module that also had the output parameter wrongly named ("output_dir" instead of "output"), unfortunately for some reason I can't understand why, despite this parameter being in the description file and despite being parsed by QGIS, it never shows in the actual GRASS command so it always fail, example:

```
Input parameters:
{ '-r' : False, 'GRASS_REGION_CELLSIZE_PARAMETER' : 0, 'GRASS_REGION_PARAMETER' : None, 'base' : '/home/giovanni/base.tif', 'bins' : 1000, 'cover' : '/home/giovanni/cover.tif', 'output' : '/tmp/processing_c1cc14a49c11459899c2d9ac5dabaa11/f9140c3da34745e29ab9c26ea1cc1f05/output', 'percentiles' : '', 'quantiles' : None }
```

`r.stats.quantile base=rast_5bc9c5092dbf04 cover=rast_5bc9c5092dc425 bins=1000 --overwrite`

(notice the lack of the "output" parameter in the GRASS command).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by gioman** [comment]
> t also "fixes" the "r.stats.quantile.rast" module that also had the output parameter wrongly named ("output_dir" instead of "output"), unfortunately for some reason I can't understand despite this parameter being in the description file and despite being parsed by QGIS it never shows in the actual GRASS command so it always fail, example:

@nyalldawson @alexbruy  any idea?

**Comment by PedroVenancio** [comment]
Check the r_stats_quantile_rast.py @gioman:

https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/grass7/ext/r_stats_quantile_rast.py

But I think the line 57 should bem changed to: 

    fileName = os.path.join(outputDir, output + '.tif')

**Comment by luipir** [comment]
> But I think the line 57 should bem changed to:
> 
> ```
> fileName = os.path.join(outputDir, output + '.tif')
> ```

you shouldn't lock to a fixed output format

**Comment by luipir** [comment]
> But I think the line 57 should bem changed to:
> 
> ```
> fileName = os.path.join(outputDir, output + '.tif')
> ```

@PedroVenancio you shouldn't lock to a fixed output format

**Comment by luipir** [comment]
> But I think the line 57 should bem changed to:
> 
> ```
> fileName = os.path.join(outputDir, output + '.tif')
> ```

@PedroVenancio you shouldn't lock to a fixed output format

> Check the r_stats_quantile_rast.py @gioman:
> 
> https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/grass7/ext/r_stats_quantile_rast.py
> 
> But I think the line 57 should bem changed to:
> 
> ```
> fileName = os.path.join(outputDir, output + '.tif')
> ```

@PedroVenancio would be better to avoid to lock to a fixed output format

**Comment by luipir** [comment]
> But I think the line 57 should bem changed to:
> 
> ```
> fileName = os.path.join(outputDir, output + '.tif')
> ```

@PedroVenancio you shouldn't lock to a fixed output format

**Comment by PedroVenancio** [comment]
Hi @luipir

I know that it would be better not to lock to a fixed format, but at this moment, the output of this tool is like this:
output_0
output_0.aux.xml
output_0.tfw
output_1
output_1.aux.xml
output_1.tfw
(...)

gdalinfo output_0

```
Driver: GTiff/GeoTIFF
Files: output_0
       output_0.aux.xml
Size is 1639, 1677
Coordinate System is:
PROJCS["unnamed",
    GEOGCS["grs80",
        DATUM["unknown",
            SPHEROID["Geodetic_Reference_System_1980",6378137,298.257222101],
            TOWGS84[0,0,0,0,0,0,0]],
        PRIMEM["Greenwich",0],
        UNIT["degree",0.0174532925199433]],
    PROJECTION["Transverse_Mercator"],
    PARAMETER["latitude_of_origin",39.66825833333333],
    PARAMETER["central_meridian",-8.133108333333334],
    PARAMETER["scale_factor",1],
    PARAMETER["false_easting",0],
    PARAMETER["false_northing",0],
    UNIT["metre",1,
        AUTHORITY["EPSG","9001"]]]
Origin = (44214.460735809996550,88315.696620079994318)
Pixel Size = (9.999799364502747,-9.999799364496118)
Metadata:
  AREA_OR_POINT=Area
  TIFFTAG_SOFTWARE=GRASS GIS 7.4.1 with GDAL 2.2.4
Image Structure Metadata:
  COMPRESSION=LZW
  INTERLEAVE=BAND
Corner Coordinates:
Upper Left  (   44214.461,   88315.697) (  7d36'42.50"W, 40d27'44.86"N)
Lower Left  (   44214.461,   71546.033) (  7d36'46.68"W, 40d18'41.21"N)
Upper Right (   60604.132,   88315.697) (  7d25' 6.88"W, 40d27'41.15"N)
Lower Right (   60604.132,   71546.033) (  7d25'12.62"W, 40d18'37.51"N)
Center      (   52409.296,   79930.865) (  7d30'57.17"W, 40d23'11.33"N)
Band 1 Block=1639x1 Type=Float32, ColorInterp=Gray
  NoData Value=nan
<GDALRasterAttributeTable>
  <FieldDefn index="0">
    <Name>min</Name>
    <Type>1</Type>
    <Usage>3</Usage>
  </FieldDefn>
  <FieldDefn index="1">
    <Name>max</Name>
    <Type>1</Type>
    <Usage>4</Usage>
  </FieldDefn>
  <FieldDefn index="2">
    <Name>redmin</Name>
    <Type>0</Type>
    <Usage>10</Usage>
  </FieldDefn>
  <FieldDefn index="3">
    <Name>redmax</Name>
    <Type>0</Type>
    <Usage>14</Usage>
  </FieldDefn>
  <FieldDefn index="4">
    <Name>greenmin</Name>
    <Type>0</Type>
    <Usage>11</Usage>
  </FieldDefn>
  <FieldDefn index="5">
    <Name>greenmax</Name>
    <Type>0</Type>
    <Usage>15</Usage>
  </FieldDefn>
  <FieldDefn index="6">
    <Name>bluemin</Name>
    <Type>0</Type>
    <Usage>12</Usage>
  </FieldDefn>
  <FieldDefn index="7">
    <Name>bluemax</Name>
    <Type>0</Type>
    <Usage>16</Usage>
  </FieldDefn>
  <Row index="0">
    <F>2</F>
    <F>277</F>
    <F>0</F>
    <F>255</F>
    <F>0</F>
    <F>255</F>
    <F>0</F>
    <F>255</F>
  </Row>
</GDALRasterAttributeTable>
```

In other words, the outputs are GeoTIFF files, but without explicit extension, what can cause confusion in less experienced users.

So, I think it would be better to declare the extension by default.

**Comment by luipir** [comment]
AFAIK any processing provider could define in an overrided method the default output format
https://qgis.org/api/classQgsProcessingProvider.html#a6438e88400b5d1969bca1ca7c10aede3
I don't know well the GRASS provider if setting this override can affect the default output format

**Comment by PedroVenancio** [comment]
@luipir 

I think when the output is a file, it works like you are saying, leaving the option to choose the file format.

But in this case, the output is a folder (QgsProcessingParameterFolderDestination) and so it does not allow the user to choose the output format. 

And because it is not hardcoded, the output file came without any extension. 

This is also a special case, where the description file needs the ext/r_stats_quantile_rast.py

This is what seems to me, but I'm not 100% certain of this.

**Comment by PedroVenancio** [comment]
Hi @luipir

I know that it would be better not to lock to a fixed format, but at this
moment, the output of this tool is like this:
output_0
output_0.aux.xml
output_0.tfw
output_1
output_1.aux.xml
output_1.tfw
(...)

gdalinfo output_0

Driver: GTiff/GeoTIFF
Files: output_0
       output_0.aux.xml
Size is 1639, 1677
Coordinate System is:
PROJCS["unnamed",
    GEOGCS["grs80",
        DATUM["unknown",

SPHEROID["Geodetic_Reference_System_1980",6378137,298.257222101],
            TOWGS84[0,0,0,0,0,0,0]],
        PRIMEM["Greenwich",0],
        UNIT["degree",0.0174532925199433]],
    PROJECTION["Transverse_Mercator"],
    PARAMETER["latitude_of_origin",39.66825833333333],
    PARAMETER["central_meridian",-8.133108333333334],
    PARAMETER["scale_factor",1],
    PARAMETER["false_easting",0],
    PARAMETER["false_northing",0],
    UNIT["metre",1,
        AUTHORITY["EPSG","9001"]]]
Origin = (44214.460735809996550,88315.696620079994318)
Pixel Size = (9.999799364502747,-9.999799364496118)
Metadata:
  AREA_OR_POINT=Area
  TIFFTAG_SOFTWARE=GRASS GIS 7.4.1 with GDAL 2.2.4
Image Structure Metadata:
  COMPRESSION=LZW
  INTERLEAVE=BAND
Corner Coordinates:
Upper Left  (   44214.461,   88315.697) (  7d36'42.50"W, 40d27'44.86"N)
Lower Left  (   44214.461,   71546.033) (  7d36'46.68"W, 40d18'41.21"N)
Upper Right (   60604.132,   88315.697) (  7d25' 6.88"W, 40d27'41.15"N)
Lower Right (   60604.132,   71546.033) (  7d25'12.62"W, 40d18'37.51"N)
Center      (   52409.296,   79930.865) (  7d30'57.17"W, 40d23'11.33"N)
Band 1 Block=1639x1 Type=Float32, ColorInterp=Gray
  NoData Value=nan
<GDALRasterAttributeTable>
  <FieldDefn index="0">
    <Name>min</Name>
    <Type>1</Type>
    <Usage>3</Usage>
  </FieldDefn>
  <FieldDefn index="1">
    <Name>max</Name>
    <Type>1</Type>
    <Usage>4</Usage>
  </FieldDefn>
  <FieldDefn index="2">
    <Name>redmin</Name>
    <Type>0</Type>
    <Usage>10</Usage>
  </FieldDefn>
  <FieldDefn index="3">
    <Name>redmax</Name>
    <Type>0</Type>
    <Usage>14</Usage>
  </FieldDefn>
  <FieldDefn index="4">
    <Name>greenmin</Name>
    <Type>0</Type>
    <Usage>11</Usage>
  </FieldDefn>
  <FieldDefn index="5">
    <Name>greenmax</Name>
    <Type>0</Type>
    <Usage>15</Usage>
  </FieldDefn>
  <FieldDefn index="6">
    <Name>bluemin</Name>
    <Type>0</Type>
    <Usage>12</Usage>
  </FieldDefn>
  <FieldDefn index="7">
    <Name>bluemax</Name>
    <Type>0</Type>
    <Usage>16</Usage>
  </FieldDefn>
  <Row index="0">
    <F>2</F>
    <F>277</F>
    <F>0</F>
    <F>255</F>
    <F>0</F>
    <F>255</F>
    <F>0</F>
    <F>255</F>
  </Row>
</GDALRasterAttributeTable>

In other words, the outputs are GeoTIFF files, but without explicit
extension, what can cause confusion in less experienced users.

So, I think it would be better to declare the extension by default.



Luigi Pirelli <notifications@github.com> escreveu no dia segunda,
22/10/2018 à(s) 08:18:

**Comment by luipir** [comment]
for me ok to merge... any objection @nyalldawson @alexbruy ?

**Review by luipir — APPROVED** [review]

**Comment by nyalldawson** [comment]
I'm not familiar with this algorithm, but the changes are self contained so I'm +1 to merge.

Of course, without a unit test, this WILL break in future. ;)
